src/ngx_http_lua_socket_tcp.c

```diff
@@ -6395,4 +6392,112 @@ ngx_http_lua_ffi_socket_tcp_del_udata(ngx_http_lua_socket_tcp_upstream_t *u,
 }
 
 
+int
+ngx_http_lua_ffi_socket_tcp_getoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    int option, int *val, u_char *err, size_t *errlen)
+{
+    u_char    errstr[NGX_MAX_ERROR_STR];
+    socklen_t len;
+    u_char   *p;
+    int       fd, r;
+
+    if (u == NULL || u->peer.connection == NULL) {
+        *errlen = ngx_snprintf(err, *errlen, "closed") - err;
+        return NGX_ERROR;
+    }
+
+    fd = u->peer.connection->fd;
+    len = sizeof(int);
+
+    switch (option) {
+    case OPTION_NAME_KEEPALIVE:
+        r = getsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, (void *) val, &len);
+        break;
+
+    case OPTION_NAME_REUSEADDR:
+        r = getsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (void *) val, &len);
+        break;
+
+    case OPTION_NAME_TCP_NODELAY:
+        r = getsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (void *) val, &len);
+        break;
+
+    case OPTION_NAME_SNDBUF:
+        r = getsockopt(fd, SOL_SOCKET, SO_RCVBUF, (void *) val, &len);
+        break;
+
+    case OPTION_NAME_RCVBUF:
+        r = getsockopt(fd, SOL_SOCKET, SO_SNDBUF, (void *) val, &len);
+        break;
+
+    default:
+        *errlen = ngx_snprintf(err, *errlen, "unsupported option") - err;
+        return NGX_ERROR;
+    }
+
+    if (r == -1) {
+        p = ngx_strerror(ngx_errno, errstr, sizeof(errstr));
+        *p = '\0';
+        *errlen = ngx_snprintf(err, *errlen, "%s", errstr) - err;
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_socket_tcp_setoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    int option, int val, u_char *err, size_t *errlen)
+{
+    u_char    errstr[NGX_MAX_ERROR_STR];
+    socklen_t len;
+    u_char   *p;
+    int       fd, r;
+
+    if (u == NULL || u->peer.connection == NULL) {
+        *errlen = ngx_snprintf(err, *errlen, "closed") - err;
+        return NGX_ERROR;
+    }
+
+    len = sizeof(int);
+    fd = u->peer.connection->fd;
+
+    switch (option) {
+    case OPTION_NAME_KEEPALIVE:
+        r = setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, (const void *) &val, len);
+        break;
+
+    case OPTION_NAME_REUSEADDR:
+        r = setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (const void *) &val, len);
+        break;
+
+    case OPTION_NAME_TCP_NODELAY:
+        r = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (const void *) &val, len);
+        break;
+
+    case OPTION_NAME_SNDBUF:
+        r = setsockopt(fd, SOL_SOCKET, SO_RCVBUF, (const void *) &val, len);
+        break;
+
+    case OPTION_NAME_RCVBUF:
+        r = setsockopt(fd, SOL_SOCKET, SO_SNDBUF, (const void *) &val, len);
+        break;
+
+    default:
+        *errlen = ngx_snprintf(err, *errlen, "unsupported option") - err;
+        return NGX_ERROR;
+    }
+
+    if (r == -1) {
+        p = ngx_strerror(ngx_errno, errstr, sizeof(errstr));
+        *p = '\0';
+        *errlen = ngx_snprintf(err, *errlen, "%s", errstr) - err;
```

Comment:
You could have used `%*s` here instead so that you don't need the trailing `\0` at all.

Reply:
modify